File: src/scraper/episodes.js
const axios = require('axios');
const cheerio = require('cheerio');
const { USER_AGENT, REQUEST_TIMEOUT } = require('../config/config');
const { getRandomUserAgent, extractEpisodeNumber } = require('../utils/helpers');

/**
 * Scrape the episode list from an anime series page
 * @param {string} url - The URL of the anime series page
 * @returns {Promise<Object>} Object containing anime title, episodes, metadata, and description
 */
const scrapeEpisodeList = async (url) => {
  try {
    // Set headers to mimic a browser request
    const headers = {
      'User-Agent': getRandomUserAgent() || USER_AGENT
    };
    
    // Send HTTP request
    const response = await axios.get(url, {
      headers,
      timeout: REQUEST_TIMEOUT
    });
    
    // Parse HTML content
    const $ = cheerio.load(response.data);
    
    // Get anime title
    const titleElement = $('h1');
    const title = titleElement.length ? titleElement.text().trim() : "Title not found";
    
    // Find episode list - check different possible containers
    let episodeSection = null;
    
    // Try different possible containers for episode lists
    const possibleContainers = [
      $('div.singlelink'),
      $('div.hq').next('div'),
      $('ul.lcp_catlist'),
      $('div.episodes')
    ];
    
    for (const container of possibleContainers) {
      if (container && container.length) {
        episodeSection = container;
        break;
      }
    }
    
    const episodes = [];
    
    if (episodeSection && episodeSection.length) {
      // Try to find episodes in list items
      const episodeItems = episodeSection.find('li');
      
      if (episodeItems.length) {
        episodeItems.each((index, item) => {
          const link = $(item).find('a');
          if (link.length) {
            const episodeTitle = link.text().trim();
            const episodeUrl = link.attr('href');
            
            // Try to extract episode number from title using regex
            const episodeNumber = extractEpisodeNumber(episodeTitle);
            
            episodes.push({
              title: episodeTitle,
              number: episodeNumber,
              url: episodeUrl
            });
          }
        });
      } else {
        // If no list items found, try to find direct links
        const links = episodeSection.find('a');
        links.each((i, link) => {
          const episodeTitle = $(link).text().trim();
          const episodeUrl = $(link).attr('href');
          
          // Try to extract episode number from title using regex
          const episodeNumber = extractEpisodeNumber(episodeTitle);
          
          episodes.push({
            title: episodeTitle,
            number: episodeNumber,
            url: episodeUrl
          });
        });
      }
    }
    
    // Sort episodes by number if possible
    episodes.sort((a, b) => {
      // Parse episode numbers to integers for comparison
      const numA = a.number === "Unknown" ? Infinity : parseInt(a.number);
      const numB = b.number === "Unknown" ? Infinity : parseInt(b.number);
      
      // Handle non-numeric values
      if (isNaN(numA) && isNaN(numB)) return 0;
      if (isNaN(numA)) return 1;
      if (isNaN(numB)) return -1;
      
      return numA - numB;
    });
    
    // Get anime metadata
    const metadata = {};
    const infoTable = $('table');
    if (infoTable.length) {
      const rows = infoTable.find('tr');
      rows.each((index, row) => {
        const header = $(row).find('th');
        const data = $(row).find('td');
        if (header.length && data.length) {
          const key = header.text().trim();
          const value = data.text().trim();
          metadata[key] = value;
        }
      });
    }
    
    // Get anime description
    let description = "";
    const descElem = $('div.unduhan');
    if (descElem.length) {
      description = descElem.text().trim();
    }
    
    return {
      title,
      episodes,
      metadata,
      description
    };
    
  } catch (error) {
    console.error(`Error fetching the episode list: ${error.message}`);
    throw new Error(`Failed to scrape episode list: ${error.message}`);
  }
};

module.exports = {
  scrapeEpisodeList
};

File: scraper/episodes.go
package scraper

import (
	"fmt"
	"net/http"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/Pendetot/AnimekApi/utils"
)

// Episode represents a single episode
type Episode struct {
	Title  string `json:"title"`
	Number string `json:"number"`
	URL    string `json:"url"`
}

// EpisodeList represents the episode list response
type EpisodeList struct {
	Title       string            `json:"title"`
	Episodes    []Episode         `json:"episodes"`
	Metadata    map[string]string `json:"metadata"`
	Description string            `json:"description"`
}

// extractEpisodeNumber extracts episode number from title using regex
func extractEpisodeNumber(title string) string {
	// Common patterns for episode numbers
	patterns := []string{
		`[Ee]pisode\s*(\d+)`,
		`[Ee]p\s*(\d+)`,
		`(\d+)`,
	}

	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		matches := re.FindStringSubmatch(title)
		if len(matches) > 1 {
			return matches[1]
		}
	}

	return "Unknown"
}

// ScrapeEpisodes scrapes the episode list from an anime series page
func ScrapeEpisodes(url string) (*EpisodeList, error) {
	// Create HTTP client with timeout
	client := &http.Client{
		Timeout: 10 * time.Second,
	}

	// Create request with headers
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %v", err)
	}

	req.Header.Set("User-Agent", utils.GetUserAgent())

	// Send request
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch page: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("received status code: %d", resp.StatusCode)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to parse HTML: %v", err)
	}

	result := &EpisodeList{
		Metadata: make(map[string]string),
	}

	// Get anime title
	titleElement := doc.Find("h1").First()
	if titleElement.Length() > 0 {
		result.Title = strings.TrimSpace(titleElement.Text())
	} else {
		result.Title = "Title not found"
	}

	// Find episode list - check different possible containers
	var episodeSection *goquery.Selection

	// Try different possible containers for episode lists
	possibleContainers := []string{
		"div.singlelink",
		"div.hq",
		"ul.lcp_catlist",
		"div.episodes",
	}

	for _, selector := range possibleContainers {
		container := doc.Find(selector)
		if container.Length() > 0 {
			episodeSection = container
			break
		}
	}

	var episodes []Episode

	if episodeSection != nil && episodeSection.Length() > 0 {
		// Try to find episodes in list items
		episodeItems := episodeSection.Find("li")

		if episodeItems.Length() > 0 {
			episodeItems.Each(func(i int, item *goquery.Selection) {
				link := item.Find("a").First()
				if link.Length() > 0 {
					episodeTitle := strings.TrimSpace(link.Text())
					if episodeUrl, exists := link.Attr("href"); exists && episodeTitle != "" {
						episodeNumber := extractEpisodeNumber(episodeTitle)

						episodes = append(episodes, Episode{
							Title:  episodeTitle,
							Number: episodeNumber,
							URL:    episodeUrl,
						})
					}
				}
			})
		} else {
			// If no list items found, try to find direct links
			links := episodeSection.Find("a")
			links.Each(func(i int, link *goquery.Selection) {
				episodeTitle := strings.TrimSpace(link.Text())
				if episodeUrl, exists := link.Attr("href"); exists && episodeTitle != "" {
					episodeNumber := extractEpisodeNumber(episodeTitle)

					episodes = append(episodes, Episode{
						Title:  episodeTitle,
						Number: episodeNumber,
						URL:    episodeUrl,
					})
				}
			})
		}
	}

	// Sort episodes by number if possible
	sort.Slice(episodes, func(i, j int) bool {
		numA, errA := strconv.Atoi(episodes[i].Number)
		numB, errB := strconv.Atoi(episodes[j].Number)

		// If both are numbers, compare numerically
		if errA == nil && errB == nil {
			return numA < numB
		}

		// If one is "Unknown", put it at the end
		if episodes[i].Number == "Unknown" {
			return false
		}
		if episodes[j].Number == "Unknown" {
			return true
		}

		// If both are non-numeric, compare as strings
		return episodes[i].Number < episodes[j].Number
	})

	result.Episodes = episodes

	// Get anime metadata
	infoTable := doc.Find("table")
	if infoTable.Length() > 0 {
		rows := infoTable.Find("tr")
		rows.Each(func(i int, row *goquery.Selection) {
			header := row.Find("th").First()
			data := row.Find("td").First()
			if header.Length() > 0 && data.Length() > 0 {
				key := strings.TrimSpace(header.Text())
				value := strings.TrimSpace(data.Text())
				result.Metadata[key] = value
			}
		})
	}

	// Get anime description
	descElem := doc.Find("div.unduhan").First()
	if descElem.Length() > 0 {
		result.Description = strings.TrimSpace(descElem.Text())
	}

	return result, nil
}
